scapy/contrib/automotive/someip.py

```diff
@@ -221,3 +235,310 @@ def _bind_someip_layers():
+        BitField("n_opt_2", 0, 4),
+        ShortField("srv_id", 0),
+        ShortField("inst_id", 0),
+        ByteField("major_ver", 0),
+        X3BytesField("ttl", 0)
+    ]
+
+    def guess_payload_class(self, payload):
+        pl_type = orb(payload[_SDEntry.TYPE_PAYLOAD_I])
+
+        if pl_type in _SDEntry.TYPE_SRV:
+            return SDEntry_Service
+        elif pl_type in _SDEntry.TYPE_EVTGRP:
+            return SDEntry_EventGroup
+
+
+class SDEntry_Service(_SDEntry):
+    _defaults = {"type": _SDEntry.TYPE_SRV_FINDSERVICE}
+
+    name = "Service Entry"
+    fields_desc = [
+        _SDEntry,
+        IntField("minor_ver", 0)
+    ]
+
+
+class SDEntry_EventGroup(_SDEntry):
+    _defaults = {"type": _SDEntry.TYPE_EVTGRP_SUBSCRIBE}
+
+    name = "Eventgroup Entry"
+    fields_desc = [
+        _SDEntry,
+        BitField("res", 0, 12),
+        BitField("cnt", 0, 4),
+        ShortField("eventgroup_id", 0)
+    ]
+
+
+# SD Option
+#  - Configuration
+#  - LoadBalancing
+# - IPv4 EndPoint
+# - IPv6 EndPoint
+
+# - IPv4 MultiCast
+# - IPv6 MultiCast
+# - IPv4 EndPoint
+# - IPv6 EndPoint
+class _SDOption(_SDPacketBase):
+    CFG_TYPE = 0x01
+    CFG_OVERALL_LEN = 4
+    LOADBALANCE_TYPE = 0x02
+    LOADBALANCE_LEN = 0x05
+    LOADBALANCE_OVERALL_LEN = 8
+    IP4_ENDPOINT_TYPE = 0x04
+    IP4_ENDPOINT_LEN = 0x0009
+    IP4_MCAST_TYPE = 0x14
+    IP4_MCAST_LEN = 0x0009
+    IP4_SDENDPOINT_TYPE = 0x24
+    IP4_SDENDPOINT_LEN = 0x0009
+    IP4_OVERALL_LEN = 12
+    IP6_ENDPOINT_TYPE = 0x06
+    IP6_ENDPOINT_LEN = 0x0015
+    IP6_MCAST_TYPE = 0x16
+    IP6_MCAST_LEN = 0x0015
+    IP6_SDENDPOINT_TYPE = 0x26
+    IP6_SDENDPOINT_LEN = 0x0015
+    IP6_OVERALL_LEN = 24
+
+    def guess_payload_class(self, payload):
+        pl_type = orb(payload[2])
+
+        return {
+            _SDOption.CFG_TYPE: SDOption_Config,
+            self.LOADBALANCE_TYPE: SDOption_LoadBalance,
+            self.IP4_ENDPOINT_TYPE: SDOption_IP4_EndPoint,
+            self.IP4_MCAST_TYPE: SDOption_IP4_Multicast,
+            self.IP4_SDENDPOINT_TYPE: SDOption_IP4_SD_EndPoint,
+            self.IP6_ENDPOINT_TYPE: SDOption_IP6_EndPoint,
+            self.IP6_MCAST_TYPE: SDOption_IP6_Multicast,
+            self.IP6_SDENDPOINT_TYPE: SDOption_IP6_SD_EndPoint
+        }.get(pl_type, Raw)
+
+
+class _SDOption_Header(_SDOption):
+    fields_desc = [
+        ShortField("len", None),
+        ByteField("type", 0),
+        ByteField("res_hdr", 0)
+    ]
+
+
+class _SDOption_Tail(_SDOption):
+    fields_desc = [
+        ByteField("res_tail", 0),
+        ByteEnumField("l4_proto", 0x06, {0x06: "TCP", 0x11: "UDP"}),
+        ShortField("port", 0)
+    ]
+
+
+class _SDOption_IP4(_SDOption):
+    fields_desc = [
+        _SDOption_Header,
+        IPField("addr", "0.0.0.0"),
+        _SDOption_Tail
+    ]
+
+
+class _SDOption_IP6(_SDOption):
+    fields_desc = [
+        _SDOption_Header,
+        IP6Field("addr", "2001:cdba:0000:0000:0000:0000:3257:9652"),
```

Comment:
Is this address specific to SD? If not, I will suggest to use `2001:db8::1` instead.

Reply:
I didn't changed any classes of someip-sd. But this looks good.